common/scala/src/main/scala/whisk/connector/lean/LeanConsumer.scala

```diff
@@ -0,0 +1,44 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package whisk.connector.lean
+
+import scala.concurrent.duration._
+import whisk.common.Logging
+import whisk.core.connector.MessageConsumer
+import java.util.concurrent.BlockingQueue
+import java.util.concurrent.TimeUnit
+
+class LeanConsumer(queue: BlockingQueue[Array[Byte]], override val maxPeek: Int)(implicit logging: Logging)
+    extends MessageConsumer {
+
+  /**
+   */
+  override def peek(duration: FiniteDuration, retry: Int): Iterable[(String, Int, Long, Array[Byte])] = {
```

Comment:
Done. On the second thought i'm not sure that matching Kafka is a good solution looking forward. For example, the recent discussion initiated by @markusthoemmes suggests that there may not be Kafka in the future architecture at all. Actually, in LeanOpenWhisk we removed Kafka which resulted in performance improvement. Adding a timeout because of the asynchronous nature of drainTo() punishes performance KPIs at high load.

Reply:
@kpavel  Not sure why do you need to sleep there. If drainTo results in an empty list then its ok and just indicates that there is nothing to drain. Next peek call would then fetch again.

From my past usage drainTo has been found to be more efficient as poll would require locking per entry while in drainTo lock is only required once for all entries present in queue at that moment. So irrespective of Kafka being used or not it would still be useful.

Reply:
@chetanmeh Without sleep it becomes a busy wait, presenting heavy impact on cpu usage. While testing locally, idle controller container shows about 200-300% cpu usage in case of using drainTo() without sleep.  It shows about 1% cpu usage when using poll().

Reply:
@kpavel That helps. I missed on fact that `peek` is designed for long poll. In such a case you previous poll based approach would then be better. Looks like default java support does not have [drainTo with wait][1]. Guava seems to provide [one][2]

[1]: https://stackoverflow.com/questions/831834/blockingqueue-blocked-drainto-methods
[2]: https://google.github.io/guava/releases/snapshot-jre/api/docs/com/google/common/collect/Queues.html#drain-java.util.concurrent.BlockingQueue-java.util.Collection-int-long-java.util.concurrent.TimeUnit-